### ISSUE TITLE
Error for kwargs annotated with complex type following optional parameters

### ISSUE THREAD

**Issue by mmarseu** [microsoft/pyright#10653]
**Describe the bug**
It seems that pyright treats optional function parameters followed by a `kwargs` with a complex type as required arguments. `kwargs` annotated with a simple type (e.g., `bool` in the minimal repro below) work as expected.

**Code**

```python
class Foo:
    pass


def works(a: int = 0, **kwargs: bool):
    pass


def doesnt_work(a: int = 0, **kwargs: Foo):
    pass


works(**{"test": False})
doesnt_work(**{"test": Foo()}) # produces error: Argument of type "Foo" cannot be assigned to parameter "a" of type "int" in function "doesnt_work" - "Foo" is not assignable to "int" (reportArgumentType)
```

**VS Code extension or command-line**
CLI: `pyright 1.1.402`

**Comment by erictraut** [comment]
This behavior is intended, although I can understand why it might look incorrect. 

This issue is related to a [recent discussion](https://discuss.python.org/t/standardizing-argument-unpacking-behaviors-for-calls/80747) that I initiated in the public Python typing forum.

I won't repeat all of the details in that thread, but to summarize... When using an unpack operator (either a single `*` or a double `**`) within a call expression, there are cases that are ambiguous at static analysis time. A static type checker needs to make assumptions that will lead to potential false positives or false negatives under certain circumstances.

Pyright's current behavior (largely) matches that of mypy. If you run mypy on your code sample above, you'll see that it produces the same results. 

The reason `works` doesn't result in an error is because `bool` is assignable to `int` whereas `Foo` is not.

One could make a good argument that mypy's and pyright's current behavior are inconsistent and should change. That's the argument I tried to make in the thread. But it's not clear whether type checkers should prefer false positives or false negatives in this case. Feel free to add to that discussion if you have an opinion one way or another.

I'm going to close this issue for now because pyright is working as designed here, and it's consistent with other type checkers. If and when we achieve consensus on a standard behavior that differs from the current behavior and the new behavior is incorporated into the Python typing spec, I will change pyright accordingly.

**Comment by mmarseu** [comment]
I see, thank you for the clarification. I guess I'd be a little out of my depth in that discussion - I fall squarely in the "casual users" category you cited in your discussion. I trust the experts to find the best solution for the most people.
